### PR TITLE
Add support for Any, All operations to Tensor

### PR DESCRIPTION
**Pull request by ashdtu**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

#1341 

### Changes

* Add support for any() (logical_or), all() (logical_and) operations on Tensor struct for all types (Float, Int, Bool). 
* For numeric tensor type(Float, Int), the input tensors are casted to Bool by checking against non-zero values. 
* The implementation supports working with these operations along a dim using the any_dim(), all_dim() methods.  
*  Burn book updated with these Tensor ops. 

### Testing

Test cases written for the implemented methods (any(), any_dim(), all(), all_dim()) with all tensor types(float, int, bool) in burn-tensor/src/tests/ops/any.rs or all.rs.

### CODE REVIEW

**Comment by ashdtu** [comment]
@louisfd The bug in WGPU about the boolean casting to float/int is no longer a blocker in the updated code. I noticed that I can get rid of that by just changing the logic a bit and adding another bool_not op which converts the weird large values of 1065353216 for (1.0 f / 1 ) to exact zeros. There's no overflow now since this op happens before the final sum, so it's all ok. 

TL;DR There's no blocker for this PR and it works for all backends.

File: crates/burn-tensor/src/tensor/ops/bool_tensor.rs [hunk 1]
```diff
@@ -302,4 +302,71 @@ pub trait BoolTensorOps<B: Backend> {
...
+    ///
+    /// # Arguments
+    ///
+    /// * `tensor` - The tensor to test.
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor with a single element, True if any element in the tensor is True, False otherwise.
+    fn bool_any<const D: usize>(tensor: BoolTensor<B, D>) -> BoolTensor<B, 1> {
+        let sum = B::int_sum(B::bool_into_int(tensor.clone()));
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/bool_tensor.rs:316]: You shouldn't need to clone tensor

File: crates/burn-tensor/src/tensor/ops/bool_tensor.rs [hunk 2]
```diff
@@ -302,4 +302,71 @@ pub trait BoolTensorOps<B: Backend> {
...
+    /// * `dim` - The axis along which to test.
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor `Tensor<B, D, Bool>` with the same size as input `tensor`, except in the `dim` axis
+    /// where the size is 1. The elem in the `dim` axis is True if any element along this dim in the input
+    /// evaluates to True, False otherwise.
+
+    fn bool_any_dim<const D: usize>(tensor: BoolTensor<B, D>, dim: usize) -> BoolTensor<B, D> {
+        let sum = B::int_sum_dim(B::bool_into_int(tensor.clone()), dim);
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/bool_tensor.rs:334]: Yous shouldn't need to clone tensor

File: crates/burn-tensor/src/tensor/ops/bool_tensor.rs [hunk 3]
```diff
@@ -302,4 +302,71 @@ pub trait BoolTensorOps<B: Backend> {
...
+    ///
+    /// * `tensor` - The tensor to test.
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor `Tensor<B, 1, Bool>` with a single element, True if all elements in the input tensor
+    /// evaluate to True, False otherwise.
+    fn bool_all<const D: usize>(tensor: BoolTensor<B, D>) -> BoolTensor<B, 1> {
+        let num_elems = B::bool_shape(&tensor).num_elements();
+        let sum = B::int_sum(B::bool_into_int(tensor.clone()));
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/bool_tensor.rs:350]: you shouldn't need to clone tensor

File: crates/burn-tensor/src/tensor/ops/bool_tensor.rs [hunk 4]
```diff
@@ -302,4 +302,71 @@ pub trait BoolTensorOps<B: Backend> {
...
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor `Tensor<B, D, Bool>` with the same size as input `tensor`, except in the `dim` axis
+    /// where the size is 1. The elem in the `dim` axis is True if all elements along this dim in the input
+    /// evaluates to True, False otherwise.
+
+    fn bool_all_dim<const D: usize>(tensor: BoolTensor<B, D>, dim: usize) -> BoolTensor<B, D> {
+        let num_elems = B::bool_shape(&tensor).dims[dim];
+        let sum = B::int_sum_dim(B::bool_into_int(tensor.clone()), dim);
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/bool_tensor.rs:369]: you shouldn't need to clone tensor

File: crates/burn-tensor/src/tensor/ops/int_tensor.rs [hunk 5]
```diff
@@ -997,4 +997,80 @@ pub trait IntTensorOps<B: Backend> {
...
+    ///
+    /// # Arguments
+    ///
+    /// * `tensor` - The tensor to test.
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor with a single element, True if any element in the tensor is True, False otherwise.
+    fn int_any<const D: usize>(tensor: IntTensor<B, D>) -> BoolTensor<B, 1> {
+        let bool_tensor = B::int_equal_elem(tensor.clone(), 0.elem());
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/int_tensor.rs:1011]: you shouldn't need to clone tensor

File: crates/burn-tensor/src/tensor/ops/int_tensor.rs [hunk 6]
```diff
@@ -997,4 +997,80 @@ pub trait IntTensorOps<B: Backend> {
...
+    /// * `dim` - The axis along which to test.
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor `Tensor<B, D, Bool>` with the same size as input `tensor`, except in the `dim` axis
+    /// where the size is 1. The elem in the `dim` axis is True if any element along this dim in the input
+    /// evaluates to True, False otherwise.
+
+    fn int_any_dim<const D: usize>(tensor: IntTensor<B, D>, dim: usize) -> BoolTensor<B, D> {
+        let bool_tensor = B::int_equal_elem(tensor.clone(), 0.elem());
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/int_tensor.rs:1031]: you shouldn't need to clone tensor

File: crates/burn-tensor/src/tensor/ops/int_tensor.rs [hunk 7]
```diff
@@ -997,4 +997,80 @@ pub trait IntTensorOps<B: Backend> {
...
+    /// * `tensor` - The tensor to test.
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor `Tensor<B, 1, Bool>` with a single element, True if all elements in the input tensor
+    /// evaluate to True, False otherwise.
+
+    fn int_all<const D: usize>(tensor: IntTensor<B, D>) -> BoolTensor<B, 1> {
+        let num_elems = B::int_shape(&tensor).num_elements();
+        let bool_tensor = B::int_equal_elem(tensor.clone(), 0.elem());
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/int_tensor.rs:1050]: you shouldn't need to clone tensor

File: crates/burn-tensor/src/tensor/ops/int_tensor.rs [hunk 8]
```diff
@@ -997,4 +997,80 @@ pub trait IntTensorOps<B: Backend> {
...
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor `Tensor<B, D, Bool>` with the same size as input `tensor`, except in the `dim` axis
+    /// where the size is 1. The elem in the `dim` axis is True if all elements along this dim in the input
+    /// evaluates to True, False otherwise.
+
+    fn int_all_dim<const D: usize>(tensor: IntTensor<B, D>, dim: usize) -> BoolTensor<B, D> {
+        let num_elems = B::int_shape(&tensor).dims[dim];
+        let bool_tensor = B::int_equal_elem(tensor.clone(), 0.0f32.elem());
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/int_tensor.rs:1071]: you shouldn't need to clone tensor

File: crates/burn-tensor/src/tensor/ops/tensor.rs [hunk 9]
```diff
@@ -1179,4 +1179,78 @@ pub trait FloatTensorOps<B: Backend> {
...
+    ///
+    /// # Arguments
+    ///
+    /// * `tensor` - The tensor to test.
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor with a single element, True if any element in the tensor is True, False otherwise.
+    fn float_any<const D: usize>(tensor: FloatTensor<B, D>) -> BoolTensor<B, 1> {
+        let bool_tensor = B::float_equal_elem(tensor.clone(), 0.0f32.elem());
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/tensor.rs:1193]: you shouldn't need to clone tensor

File: crates/burn-tensor/src/tensor/ops/tensor.rs [hunk 10]
```diff
@@ -1179,4 +1179,78 @@ pub trait FloatTensorOps<B: Backend> {
...
+        let bool_tensor = B::bool_not(bool_tensor);
+        let sum = B::float_sum(B::bool_into_float(bool_tensor));
+        B::float_greater_elem(sum, 0.0f32.elem())
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/tensor.rs:1196]: I think `0.elem()` suffices?

File: crates/burn-tensor/src/tensor/ops/tensor.rs [hunk 11]
```diff
@@ -1179,4 +1179,78 @@ pub trait FloatTensorOps<B: Backend> {
...
+    /// * `dim` - The axis along which to test.
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor `Tensor<B, D, Bool>` with the same size as input `tensor`, except in the `dim` axis
+    /// where the size is 1. The elem in the `dim` axis is True if any element along this dim in the
+    /// input evaluates to True, False otherwise.
+
+    fn float_any_dim<const D: usize>(tensor: FloatTensor<B, D>, dim: usize) -> BoolTensor<B, D> {
+        let bool_tensor = B::float_equal_elem(tensor.clone(), 0.0f32.elem());
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/tensor.rs:1213]: you shouldn't need to clone tensor

File: crates/burn-tensor/src/tensor/ops/tensor.rs [hunk 12]
```diff
@@ -1179,4 +1179,78 @@ pub trait FloatTensorOps<B: Backend> {
...
+    ///
+    /// * `tensor` - The tensor to test.
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor `Tensor<B, 1, Bool>` with a single element, True if all elements in the input tensor
+    /// evaluate to True, False otherwise.
+    fn float_all<const D: usize>(tensor: FloatTensor<B, D>) -> BoolTensor<B, 1> {
+        let num_elems = B::float_shape(&tensor).num_elements();
+        let bool_tensor = B::float_equal_elem(tensor.clone(), 0.0f32.elem());
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/tensor.rs:1231]: shouldn't need to clone tensor

File: crates/burn-tensor/src/tensor/ops/tensor.rs [hunk 13]
```diff
@@ -1179,4 +1179,78 @@ pub trait FloatTensorOps<B: Backend> {
...
+    /// * `dim` - The axis along which to test.
+    ///
+    /// # Returns
+    ///
+    /// A boolean tensor `Tensor<B, D, Bool>` with the same size as input `tensor`, except in the `dim` axis
+    /// where the size is 1. The elem in the `dim` axis is True if all elements along this dim in the input
+    /// evaluates to True, False otherwise.
+    fn float_all_dim<const D: usize>(tensor: FloatTensor<B, D>, dim: usize) -> BoolTensor<B, D> {
+        let num_elems = B::float_shape(&tensor).dims[dim];
+        let bool_tensor = B::float_equal_elem(tensor.clone(), 0.0f32.elem());
```
Comment (louisfd) [inline crates/burn-tensor/src/tensor/ops/tensor.rs:1251]: you shouldn't need to clone tensor

File: crates/burn-tensor/src/tests/mod.rs [hunk 14]
```diff
@@ -79,6 +79,8 @@ macro_rules! testgen_all {
         burn_tensor::testgen_transpose!();
         burn_tensor::testgen_tri!();
         burn_tensor::testgen_powf!();
+        burn_tensor::testgen_any!();
+        burn_tensor::testgen_all_op!();
```
Comment (louisfd) [inline crates/burn-tensor/src/tests/mod.rs:83]: oh we had a name clash here! all_op is a good name

File: crates/burn-tensor/src/tests/ops/any.rs [hunk 15]
```diff
@@ -0,0 +1,46 @@
+#[burn_tensor_testgen::testgen(any)]
+mod tests {
+    use super::*;
+    use burn_tensor::{Data, Tensor};
+
+    #[test]
+    fn test_any() {
```
Comment (louisfd) [inline crates/burn-tensor/src/tests/ops/any.rs:7]: I think we should also have versions of this test that give false. Otherwise the op could just be `return true`

File: crates/burn-tensor/src/tests/ops/all.rs [hunk 16]
```diff
@@ -0,0 +1,46 @@
+#[burn_tensor_testgen::testgen(all_op)]
+mod tests {
+    use super::*;
+    use burn_tensor::{Data, Tensor};
+
+    #[test]
+    fn test_all() {
```
Comment (louisfd) [inline crates/burn-tensor/src/tests/ops/all.rs:7]: I think we should also have versions of this test that give true. Otherwise the op could just be `return false`

**Review by louisfd — CHANGES_REQUESTED** [review]
Awesome! Thank you. 
The logic is good everywhere. I do have some comments though, most of them minor but one is important, you must not do tensor.clone() if you're not obligated, this will augment reference count and may prevent operations to happen in-place. To make sure no occurrence of the clone is forgotten i commented them all 😄

File: crates/burn-tensor/src/tests/mod.rs (hunk 14 above)
Comment (antimora) [reply crates/burn-tensor/src/tests/mod.rs:83]: For consistency, I would have renamed `testgen_all` to something else. Maybe `testgen_every_op`?

**Comment by louisfd** [comment]
@ashdtu While we're here, can you change this LSTM test line in lstm.rs (which was the original reason I needed the any op):
```
 // Asserts the gradients exist and are non zero
 assert!(*some_gradient.abs().sum().into_data().value.first().unwrap() > 0.);
 ```

File: crates/burn-tensor/src/tests/mod.rs (hunk 14 above)
Comment (louisfd) [reply crates/burn-tensor/src/tests/mod.rs:83]: Not a bad idea, but goes beyond the scope of this PR. all_op is good for now
